website/blog/2019-12-20-announcing-sorbet-0.5.md

```diff
@@ -0,0 +1,540 @@
+variable:
+
+```ruby
+# typed: strict
+module B
+  sig {returns(String)}
+  def current_user
+    # error: Use of undeclared variable `@current_user`
+    @current_user ||= ENV.fetch('USER')
+  end
+end
+```
+
+The change we made is to allow instance variables to be declared **anywhere**,
+as long as they're declared nilable. Sorbet still guarantees that **if** these
+instance variables are initialized they have the right type, but doesn't make a
+promise about whether or not they're initialized at all.
+
+Using `typed: strict` with our `current_user` example from before now involves
+just a single line to declare the type of the `@current_user` variable:
+
+```ruby
+# typed: strict
+module B
+  sig {returns(String)}
+  def current_user
+    # Declare @current_user as either String or nil:
+    @current_user = T.let(@current_user, T.nilable(String))
+    @current_user ||= ENV.fetch('USER')
+  end
+end
+```
+
+Collectively, these three features have made `typed: strict` adoption much
+smoother, and we have the numbers to back it up. In Stripe's multi-million-line
+Ruby codebase, over 85% of files are `typed: true` or above, and over 40% of all
+files are `typed: strict` or above.
+
+That's pretty much it for the `typed: strict` features, so with the next section
+we're back to normal type system features.
+
+### Type checking database code with `T.attached_class`
+
+Sorbet has had `T.class_of(MyClass)` [since it was open sourced][class-of-docs],
+which allows passing class objects around as a values. But Sorbet lacked the
+opposite feature: to refer to "an instance of the current singleton class." This
+pattern is super common in real-world Ruby: it's how basically every ORM's API
+is structured:
+
+```ruby
+# ActiveRecord:
+Person.find(1) # => returns an *instance* of Person
+
+# Stripe's internal ORM:
+Charge.load_one('ch_some_id') # => returns an *instance* of Charge
+```
+
+It wasn't previously possible to write a return type for these methods, but it's
+simple now, via a feature we call `T.attached_class`:
+
+```ruby
+class AbstractModel
+  # Returns `T.attached_class`, or "an instance of
+  # whatever the current singleton class is"
+  sig {params(id: String).returns(T.attached_class)}
+  def self.load_one(id)
+    # ... calls self.new somewhere ...
+  end
+end
+```
+
+It's interesting to think about **why** typing these methods (`find`,
+`load_one`, ...) wasn't possible before. At first glance, it almost seems like
+these methods need a different type annotation depending on the call site. Call
+`find` on the `Person` class to get back a `Person` instance, but call it on the
+`Charge` class to get back a `Charge` instance. And while it looks like the type
+needs to be different each time, these methods are only written once (inside the
+framework itself) which means there can only be one sig that captures all these
+different behaviors.
+
+This problem—defining something once, but using it polymorphically at many
```

Comment:
"at many types" sounds wrong to me?

Reply:
This is somewhat of a technical term when talking about generics. For example, consider an identity function:

```haskell
id :: forall a. a -> a
id x = x
```

The type is `forall a. a -> a`. The `forall` here means that this function can be used _at many types_ depending on what argument `a` it's given.

```haskell
id 0 -- a = Int, so substituting into forall a. a -> a  gives  Int -> Int
```

This is because behind the scenes what the type checker is really doing is figuring out what type to call it at, and providing that type to the function explicitly, something like this:

```haskell
id @Int 0 -- @Int means "instantiate this polymorphic function at the type Int
```

So yeah it's maybe a bit jargony, but I think I'm going to keep it.